File: app/src/main/java/com/themaker/fshmo/klassikaplus/service/NetworkUtils.java
package com.themaker.fshmo.klassikaplus.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.work.WorkManager;
import com.themaker.fshmo.klassikaplus.App;
import com.themaker.fshmo.klassikaplus.presentation.root.MainActivity;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;

import java.util.UUID;

import static android.content.Context.CONNECTIVITY_SERVICE;

public class NetworkUtils {

    private static NetworkUtils networkUtils;
    private NetworkReceiver networkReceiver = new NetworkReceiver();
    private NotificationTapReceiver notificationTapReceiver = new NotificationTapReceiver();
    private Subject<Boolean> networkState = BehaviorSubject.createDefault(isNetworkAvailable());


    public static NetworkUtils getInstance() {
        synchronized (App.class) {
            if (networkUtils == null)
                synchronized (App.class) {
                    networkUtils = new NetworkUtils();
                }
            return networkUtils;
        }
    }

    private boolean isNetworkAvailable() {
        ConnectivityManager connectivityManager = (ConnectivityManager) App.getInstance()
                .getApplicationContext()
                .getSystemService(CONNECTIVITY_SERVICE);
        if (connectivityManager == null)
            return false;
        NetworkInfo info = connectivityManager.getActiveNetworkInfo();
        return info != null && info.isConnected();
    }

    public NetworkReceiver getNetworkReceiver() {
        return networkReceiver;
    }

    public Single<Boolean> getOnlineNetwork() {
        return networkState.subscribeOn(Schedulers.io())
                .filter(online -> online)
                .firstOrError();
    }

    public NotificationTapReceiver getNotificationTapReceiver() {
        return notificationTapReceiver;
    }

    public class NetworkReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            networkState.onNext(isNetworkAvailable());
        }
    }

    public class NotificationTapReceiver extends BroadcastReceiver {

        public static final String ACTION_TAP = "Notification tapped";
        private final String TAG = NotificationTapReceiver.class.getName();
        private UUID workRequestId;

        @Override
        public void onReceive(Context context, Intent intent) {
            MainActivity.start(context);
            Log.i(TAG, "Starting App");
        }

        public void setWorkRequestId(@NonNull UUID workRequestId) {
            this.workRequestId = workRequestId;
        }
    }
}

File: app/src/main/java/com/themaker/fshmo/klassikaplus/service/NetworkUtils.kt
package com.themaker.fshmo.klassikaplus.service

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.util.Log
import androidx.work.WorkManager
import com.themaker.fshmo.klassikaplus.App
import com.themaker.fshmo.klassikaplus.presentation.root.MainActivity
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.Subject

import java.util.UUID

import android.content.Context.CONNECTIVITY_SERVICE

class NetworkUtils {
    val networkReceiver = NetworkReceiver()
    val notificationTapReceiver = NotificationTapReceiver()
    private val networkState = BehaviorSubject.createDefault(isNetworkAvailable)

    private val isNetworkAvailable: Boolean
        get() {
            val connectivityManager = App.getInstance()
                .applicationContext
                .getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager ?: return false
            val info = connectivityManager.activeNetworkInfo
            return info != null && info.isConnected
        }

    val onlineNetwork: Single<Boolean>
        get() = networkState.subscribeOn(Schedulers.io())
            .filter { online -> online }
            .firstOrError()

    inner class NetworkReceiver : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            networkState.onNext(isNetworkAvailable)
        }
    }

    inner class NotificationTapReceiver : BroadcastReceiver() {
        private val TAG = NotificationTapReceiver::class.java.name
        private var workRequestId: UUID? = null
        val ACTION_TAP = "Notification tapped"

        override fun onReceive(context: Context, intent: Intent) {
            MainActivity.start(context)
            Log.i(TAG, "Starting App")
        }

        fun setWorkRequestId(workRequestId: UUID) {
            this.workRequestId = workRequestId
        }
    }

    companion object {
        private val TAG = NetworkUtils::class.java.name

        init {
            Log.i(TAG, "INIT")
        }

        fun instance(): NetworkUtils = NetworkUtils()

    }
}